src/balance.ts

```diff
@@ -1,72 +1,62 @@
-import * as errors from "./errors";
 import { BigNumber } from "bignumber.js";
+import { ESDTToken } from "./esdtToken";
+import { ErrInvalidArgument } from "./errors";
+import { Egld } from "./balanceBuilder";
 
 /**
- * The base used for toString methods to avoid exponential notation
+ * The number of decimals handled when working with EGLD or ESDT values.
  */
-const BASE_10 = 10;
+const DEFAULT_BIGNUMBER_DECIMAL_PLACES = 18;
 
-/**
- * The number of decimals handled when working with EGLD values.
- */
-const DENOMINATION = 18;
-
-/**
- * One EGLD, in its big-integer form (as a string).
- */
-const OneEGLDString = "1000000000000000000";
-
-const EGLDTicker = "EGLD";
 
-BigNumber.set({ DECIMAL_PLACES: DENOMINATION, ROUNDING_MODE: 1 });
+BigNumber.set({ DECIMAL_PLACES: DEFAULT_BIGNUMBER_DECIMAL_PLACES, ROUNDING_MODE: 1 });
 
 /**
  * Balance, as an immutable object.
- * TODO: Re-design, perhaps as new Money(value, currency), with new Currency(denomination, ticker). 
  */
 export class Balance {
+    readonly token: ESDTToken;
+    private readonly nonce: BigNumber = new BigNumber(0);
     private readonly value: BigNumber = new BigNumber(0);
 
     /**
      * Creates a Balance object.
      */
-    public constructor(value: string) {
+    public constructor(token: ESDTToken, nonce: BigNumber.Value, value: BigNumber.Value) {
+        this.token = token;
+        this.nonce = new BigNumber(nonce);
         this.value = new BigNumber(value);
-
-        if (this.value.isNegative()) {
```

Comment:
In a few examples, I wanted to check the balance difference of an account and print it.
For example, in the crowdfunding example I check if alice has received the 3 EGLD or if the funding has failed.
If the funding has failed, the resulting balance is negative (because of the gas spent).
I think representing a negative balance should be allowed, and the validation for positive values should be added on a case-by-case basis.